File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/task/TaskImpl.java
package com.example.alexeyglushkov.taskmanager.task;

import android.os.Looper;
import android.util.Log;

import com.example.alexeyglushkov.streamlib.progress.ProgressInfo;
import com.example.alexeyglushkov.streamlib.progress.ProgressListener;
import com.example.alexeyglushkov.streamlib.progress.ProgressUpdater;
import com.example.alexeyglushkov.tools.HandlerTools;

import junit.framework.Assert;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Date;

/**
 * Created by alexeyglushkov on 23.07.15.
 */
public abstract class TaskImpl implements Task, TaskPrivate {
    protected Task.Callback callback;
    protected Task.Callback startCallback;
    protected Object cancellationInfo;
    protected Task.Status taskStatus = Task.Status.NotStarted;
    protected Object taskUserData;
    protected boolean needCancelTask;
    protected boolean isCancelled; // is set in subclasses when task is really cancelled
    protected float progressMinChange = 0.1f;
    protected Error error;
    protected Object result;
    protected String taskId;
    protected LoadPolicy loadPolicy = LoadPolicy.SkipIfAdded;
    protected int priority;
    protected int type;

    protected Date startDate;
    protected Date finishDate;

    protected WeakRefList<Task> blockedByTasks;

    // listeners are cleared in a TaskManager after task finishing or cancelling
    protected ArrayList<StatusListener> statusListeners;
    protected WeakRefList<ProgressListener> progressListeners;

    public TaskImpl() {
        super();

        blockedByTasks = new WeakRefList<Task>();
        statusListeners = new ArrayList<StatusListener>();
        progressListeners = new WeakRefList<>();
    }

    @Override
    public Callback getTaskCallback() {
        return this.callback;
    }

    @Override
    public Callback getStartCallback() {
        return startCallback;
    }

    @Override
    public void setTaskCallback(Callback callback) {
        this.callback = callback;
    }

    @Override
    public void cancelTask(Object info) {
        needCancelTask = true;
        this.cancellationInfo = info;
    }

    @Override
    public void setTaskStatus(Task.Status status) {
        Task.Status oldStatus = this.taskStatus;
        this.taskStatus = status;

        Log.d("TaskImpl", "setTaskStatus " + status + " " + Thread.currentThread()); //TODO: CRASH: get concurrent access crash in triggerStatusListeners
        triggerStatusListeners(oldStatus, this.taskStatus);
    }

    @Override
    public Task.Status getTaskStatus() {
        return this.taskStatus;
    }

    public void setTaskError(Error error) {
        this.error = error;
    }

    @Override
    public Error getTaskError() {
        return error;
    }

    @Override
    public Object getTaskUserData() {
        return taskUserData;
    }

    @Override
    public void setTaskUserData(Object object) {
        this.taskUserData = object;
    }

    @Override
    public void setTaskPriority(int value) {
        priority = value;
    }

    @Override
    public int getTaskPriority() {
        return priority;
    }

    @Override
    public void setTaskId(String id) {
        this.taskId = id;
    }

    @Override
    public String getTaskId() {
        return this.taskId;
    }

    @Override
    public void setLoadPolicy(LoadPolicy loadPolicy) {
        this.loadPolicy = loadPolicy;
    }

    @Override
    public LoadPolicy getLoadPolicy() {
        return this.loadPolicy;
    }

    @Override
    public Task getTask() {
        return this;
    }

    @Override
    public void setTaskProgressMinChange(float value) {
        progressMinChange = value;
    }

    @Override
    public float getTaskProgressMinChange() {
        return progressMinChange;
    }

    @Override
    public Object getCancellationInfo() {
        return cancellationInfo;
    }

    @Override
    public void setTaskType(int type) {
        this.type = type;
    }

    @Override
    public int getTaskType() {
        return type;
    }

    @Override
    public void startTask(Callback callback) {
        getPrivate().handleTaskStart(callback);
    }

    @Override
    public boolean getNeedCancelTask() {
        return needCancelTask;
    }

    public Object getTaskResult() {
        return result;
    }

    public void setTaskResult(Object result) {
        this.result = result;
    }

    @Override
    public void addTaskStatusListener(StatusListener listener) {
        statusListeners.add(listener);
    }

    @Override
    public void removeTaskStatusListener(StatusListener listener) {
        for (int i=0; i<statusListeners.size(); ++i) {
            if (statusListeners.get(i) == listener) {
                statusListeners.remove(i);
                break;
            }
        }
    }

    @Override
    public void addTaskProgressListener(ProgressListener listener) {
        checkMainThread();

        progressListeners.add(new WeakReference<ProgressListener>(listener));
    }

    @Override
    public void removeTaskProgressListener(ProgressListener listener) {
        checkMainThread();

        int i=0;
        for (WeakReference<ProgressListener> l : progressListeners) {
            if (l.get() == listener) {
                progressListeners.remove(i);
                break;
            }

            ++i;
        }
    }

    @Override
    public void clearAllListeners() {
        statusListeners.clear();

        if (progressListeners.size() > 0) {
            HandlerTools.runOnMainThread(new Runnable() {
                @Override
                public void run() {
                    progressListeners.clear();
                }
            });
        }
    }

    @Override
    public long getTaskDuration() {
        if (startDate != null) {
            long finishTime = finishDate == null ? new Date().getTime() : finishDate.getTime();
            return finishTime - startDate.getTime();
        }
        return -1;
    }

    @Override
    public void setTaskStartDate(Date date) {
        this.startDate = date;
    }

    @Override
    public void setTaskFinishDate(Date date) {
        this.finishDate = date;
    }

    //TODO: add dependencies support

    @Override
    public void addTaskDependency(Task task) {
        blockedByTasks.add(new WeakReference<Task>(task));
    }

    @Override
    public void removeTaskDependency(Task task) {
        blockedByTasks.remove(task);
    }

    public WeakRefList<Task> getDependencies() {
        return blockedByTasks;
    }

    public boolean isBlocked() {
        boolean isBlocked = false;
        for (WeakReference<Task> task : blockedByTasks) {
            isBlocked = !Tasks.isTaskCompleted(task.get());
            if (isBlocked) {
                break;
            }
        }

        return isBlocked;
    }

    @Override
    public void clear() {
        Assert.assertTrue(Tasks.isTaskCompleted(this));

        taskStatus = Status.NotStarted;
        cancellationInfo = null;
        needCancelTask = false;
        isCancelled = false;
        error = null;
        result = null;
        startDate = null;
        finishDate = null;
    }

    @Override
    public TaskPrivate getPrivate() {
        return this;
    }

    @Override
    public ProgressUpdater createProgressUpdater(float contentSize) {
        ProgressUpdater updater = new ProgressUpdater(contentSize, getTaskProgressMinChange(), new ProgressUpdater.ProgressUpdaterListener() {
            @Override
            public void onProgressUpdated(ProgressUpdater updater) {
                triggerProgressListeners(updater);
            }

            @Override
            public void onProgressCancelled(ProgressUpdater updater, Object info) {
                triggerProgressListeners(updater);
                cancelTask(info);
            }
        });
        return updater;
    }

    private void triggerStatusListeners(Task.Status oldStatus, Task.Status newStatus) {
        if (oldStatus != newStatus) {
            for (StatusListener l : statusListeners) {
                if (l != null) {
                    l.onTaskStatusChanged(TaskImpl.this, oldStatus, newStatus);
                }
            }
        }
    }

    public void triggerProgressListeners(final ProgressInfo progressInfo) {
        if (progressListeners.size() > 0) {
            HandlerTools.runOnMainThread(new Runnable() {
                @Override
                public void run() {
                    for (WeakReference<ProgressListener> l : progressListeners) {
                        l.get().onProgressChanged(TaskImpl.this, progressInfo);
                    }
                }
            });
        }
    }

    protected void setIsCancelled() {
        isCancelled = true;
    }

    @Override
    public boolean canBeCancelledImmediately() {
        return false;
    }

    @Override
    public void handleTaskStart(Callback callback) {
        startCallback = callback;
    }

    @Override
    public void handleTaskCompletion(Callback callback) {
        if (callback != null) {
            callback.onCompleted(isCancelled);
        }
    }

    private void checkMainThread() {
        Assert.assertEquals(Looper.getMainLooper().getThread(), Thread.currentThread());
    }
}

File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/task/TaskImpl.kt
package com.example.alexeyglushkov.taskmanager.task

import android.os.Looper
import android.util.Log
import androidx.annotation.WorkerThread

import com.example.alexeyglushkov.streamlib.progress.ProgressInfo
import com.example.alexeyglushkov.streamlib.progress.ProgressListener
import com.example.alexeyglushkov.streamlib.progress.ProgressUpdater
import com.example.alexeyglushkov.taskmanager.tools.WeakRefList
import com.example.alexeyglushkov.tools.HandlerTools

import org.junit.Assert
import java.lang.Exception

import java.lang.ref.WeakReference
import java.util.Date

/**
 * Created by alexeyglushkov on 23.07.15.
 */
abstract class TaskImpl : TaskBase, TaskPrivate {
    // Input
    final override var taskId: String? = null
    final override var loadPolicy: Task.LoadPolicy = Task.LoadPolicy.SkipIfAlreadyAdded
    final override var taskPriority: Int = 0
    final override var taskType: Int = 0
    final override var taskUserData: Any? = null
    final override var finishCallback: Task.Callback? = null
    final override var cancellationInfo: Any? = null
        protected set
    final override var taskProgressMinChange = 0.1f

    override var dependencies: WeakRefList<Task> = WeakRefList()
        protected set

    // listeners are cleared in a TaskManager after task finishing or cancelling
    private var statusListeners = mutableListOf<Task.StatusListener>()
    private var progressListeners = WeakRefList<ProgressListener>()

    // Output
    override var taskResult: Any? = null // TODO: use generic type
    override var taskError: Exception? = null
    override var needCancelTask: Boolean = false
        protected set
    protected var isCancelled: Boolean = false // is set in subclasses when task is really cancelled
    private var startDate: Date? = null
    private var finishDate: Date? = null

    private var _taskStatus: Task.Status = Task.Status.NotStarted
    override var taskStatus: Task.Status
        get() = _taskStatus
        set(value) {
            log("TaskImpl", "setTaskStatus $value from $_taskStatus")

            val oldStatus = _taskStatus
            _taskStatus = value

            Log.d("TaskImpl", "setTaskStatus " + value + " " + taskId + " " + Thread.currentThread())
            triggerStatusListeners(oldStatus, this.taskStatus)
        }

    override val task: Task
        get() = this

    override val private: TaskPrivate
        get() = this

    override fun taskDuration(): Long {
        val startDate = startDate
        if (startDate != null) {
            val finishDate = finishDate
            val finishTime = finishDate?.time ?: Date().time
            return finishTime - startDate.time
        }
        return -1
    }

    override fun hasActiveDependencies(): Boolean {
        var isBlocked = false
        for (task in dependencies) {
            val realTask = task.get()
            if (realTask != null) {
                isBlocked = !realTask.isFinished()
                if (isBlocked) {
                    break
                }
            }
        }

        return isBlocked
    }

    override fun cancelTask(info: Any?) {
        needCancelTask = true
        cancellationInfo = info
    }

    @WorkerThread
    override fun addTaskStatusListener(listener: Task.StatusListener) {
        statusListeners.add(listener)
    }

    @WorkerThread
    override fun removeTaskStatusListener(listener: Task.StatusListener) {
        for (i in 0 until statusListeners.size) {
            if (statusListeners[i] === listener) {
                statusListeners.removeAt(i)
                break
            }
        }
    }

    override fun addTaskProgressListener(listener: ProgressListener) {
        progressListeners.add(WeakReference(listener))
    }

    override fun removeTaskProgressListener(listener: ProgressListener) {
        checkMainThread()

        var i = 0
        for (l in progressListeners) {
            if (l.get() === listener) {
                progressListeners.removeAt(i)
                break
            }

            ++i
        }
    }

    @WorkerThread
    override fun clearAllListeners() {
        statusListeners.clear()

        if (progressListeners.size > 0) {
            HandlerTools.runOnMainThread { progressListeners.clear() }
        }
    }

    override fun setTaskStartDate(date: Date) {
        this.startDate = date
    }

    override fun setTaskFinishDate(date: Date) {
        this.finishDate = date
    }

    override fun addTaskDependency(task: Task) {
        dependencies.add(WeakReference(task))
    }

    override fun removeTaskDependency(task: Task) {
        dependencies.removeValue(task)
    }

    override fun clear() {
        Assert.assertTrue(isFinished())

        _taskStatus = Task.Status.NotStarted
        cancellationInfo = null
        needCancelTask = false
        isCancelled = false
        taskError = null
        taskResult = null
        startDate = null
        finishDate = null
    }

    override fun createProgressUpdater(contentSize: Float): ProgressUpdater {
        return ProgressUpdater(contentSize, taskProgressMinChange, object : ProgressUpdater.ProgressUpdaterListener {
            override fun onProgressUpdated(updater: ProgressUpdater) {
                triggerProgressListeners(updater)
            }

            override fun onProgressCancelled(updater: ProgressUpdater, info: Any?) {
                triggerProgressListeners(updater)
                cancelTask(info)
            }
        })
    }

    @WorkerThread
    private fun triggerStatusListeners(oldStatus: Task.Status, newStatus: Task.Status) {
        if (oldStatus != newStatus) {
            val listeners = statusListeners.toList() // create a copy as the statusListeners might be modified in onTaskStatusChanged
            for (l in listeners) {
                l.onTaskStatusChanged(this@TaskImpl, oldStatus, newStatus)
            }
        }
    }

    override fun triggerProgressListeners(progressInfo: ProgressInfo) {
        if (progressListeners.size > 0) {
            HandlerTools.runOnMainThread {
                for (l in progressListeners) {
                    l.get()?.onProgressChanged(this@TaskImpl, progressInfo)
                }
            }
        }
    }

    protected fun setIsCancelled() {
        isCancelled = true
    }

    override fun canBeCancelledImmediately(): Boolean {
        return false
    }

    private fun checkMainThread() {
        Assert.assertEquals(Looper.getMainLooper().thread, Thread.currentThread())
    }
}
